src/DicomWebClient/CLI/Qido.cs

```diff
@@ -0,0 +1,148 @@
+            [Option("f", "fuzzy matching")] bool fuzzyMatching = false,
+            [Option("o", "output directory, default: current directory(.)", DefaultValue = ".")] string outputDir = "."
+            )
+        {
+            Uri rootUri;
+            ValidateOptions(rootUrl, out rootUri);
+            ValidateOutputDirectory(ref outputDir);
+
+            var client = new DicomWebClient(rootUri, Utils.GenerateFromUsernamePassword(username, password));
+            _logger.LogInformation($"Querying studies...");
+
+            var queryParameters = ParseQueryString(query);
+
+            await SaveJson(outputDir, client.Qido.SearchForStudies(queryParameters, fieldsToInclude, fuzzyMatching));
+        }
+
+        private Dictionary<string, string> ParseQueryString(string query)
+        {
+            var pairs = query.Split(",", StringSplitOptions.RemoveEmptyEntries);
+            var queryParameters = new Dictionary<string, string>();
+            foreach (var pair in pairs)
+            {
+                var parts = pair.Split("=", StringSplitOptions.RemoveEmptyEntries);
+                if (parts.Length != 2)
+                {
+                    _logger.LogError("Unknown query parameter {0}", pair);
+                }
+                if (queryParameters.ContainsKey(parts[0]))
+                {
+                    throw new Exception($"Query key already exists {parts[0]}");
+                }
+                queryParameters.Add(parts[0], parts[1]);
+            }
+            return queryParameters;
+        }
+
+        private async Task SaveJson(string outputDir, IAsyncEnumerable<string> enumerable)
+        {
+            Guard.Against.NullOrWhiteSpace(outputDir, nameof(outputDir));
+            Guard.Against.Null(enumerable, nameof(enumerable));
+
+            await foreach (var item in enumerable)
+            {
+                await Utils.SaveJson(_logger, outputDir, item, DicomTag.StudyInstanceUID);
+            }
+        }
+
+        private async Task SaveFiles(string outputDir, IAsyncEnumerable<DicomDataset> enumerable)
+        {
+            Guard.Against.NullOrWhiteSpace(outputDir, nameof(outputDir));
+            Guard.Against.Null(enumerable, nameof(enumerable));
+
+            var count = 0;
+            await foreach (var dataset in enumerable)
+            {
+                var file = new DicomFile(dataset);
+                var filename = file.Dataset.GetSingleValueOrDefault(DicomTag.StudyInstanceUID, $"unknown-study-{count}");
+                await Utils.SaveFiles(_logger, file, Path.Combine(outputDir, filename));
+                count++;
+            }
+            _logger.LogInformation($"Successfully saved {count} files.");
+        }
+
+        private void ValidateOutputFilename(ref string filename)
+        {
+            Guard.Against.NullOrWhiteSpace(filename, nameof(filename));
+
+            try
+            {
+                filename = Path.GetFullPath(filename);
```

Comment:
if `Path.GetFullPath` throws, it's likely because `filename` is illegal -- probably should not just squelch this as the information could be very useful to the caller

Reply:
Fixed 🙇